memtable/wbwi_memtable.cc

```diff
@@ -0,0 +1,147 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "memtable/wbwi_memtable.h"
+
+#include "db/memtable.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+const std::unordered_map<WriteType, ValueType>
+    WBWIMemTableIterator::WriteTypeToValueTypeMap = {
+        {kPutRecord, kTypeValue},
+        {kMergeRecord, kTypeMerge},
+        {kDeleteRecord, kTypeDeletion},
+        {kSingleDeleteRecord, kTypeSingleDeletion},
+        {kDeleteRangeRecord, kTypeRangeDeletion},
+        {kPutEntityRecord, kTypeWideColumnEntity},
+        // Only the above record types are added to WBWI.
+        // kLogDataRecord, kXIDRecord, kUnknownRecord
+};
+
+bool WBWIMemTable::Get(const LookupKey& key, std::string* value,
+                       PinnableWideColumns* columns, std::string* timestamp,
+                       Status* s, MergeContext* merge_context,
+                       SequenceNumber* max_covering_tombstone_seq,
+                       SequenceNumber* out_seq, const ReadOptions&,
+                       bool immutable_memtable, ReadCallback* callback,
+                       bool* is_blob_index, bool do_merge) {
+  (void)immutable_memtable;
+  (void)timestamp;
+  (void)columns;
+  assert(immutable_memtable);
+  assert(!timestamp);  // TODO: support UDT
+  assert(!columns);    // TODO: support WideColumn
+  assert(global_seqno_ != kMaxSequenceNumber);
+  // WBWI does not support DeleteRange yet.
+  assert(!wbwi_->GetWriteBatch()->HasDeleteRange());
```

Comment:
nit: Maybe add assertion that such a MemTable is never empty.

Reply:
For our usage in transaction this will be true, but I don't want to assume that inside WBWIMemTable.